Add `-S|--session` command line option

johanmalm opened the pull request:
See commit message.

Ref #1595

- [x] Documentation
- [ ] Testing

Cc: @stefonarch

ahesford commented:
I think the concept is good, but "primary-client" might be a bit opaque. What about `-S` and `--session-manager` or `--session-startup`, which might be a bit more intuitive and has the added benefit that it mirrors the `-s` short option. Then people can remember "s" for startup, and capitalizing the argument makes it "important".

stefonarch commented:
Not sure if I miss something, but
```
labwc --primary-client qterminal
Usage: labwc [options...]
  -c, --config <file>      Specify config file (with path)
  -C, --config-dir <dir>   Specify config directory
```
but thanks, this would solve a quite big issue :)

johanmalm commented:
> Not sure if I miss something

Sorry, I messed up. Just try with `labwc -p qterminal` for now.

johanmalm commented:
> I think the concept is good, but "primary-client" might be a bit opaque. What about `-S` and `--session-manager` or `--session-startup`, which might be a bit more intuitive and has the added benefit that it mirrors the `-s` short option. Then people can remember "s" for startup, and capitalizing the argument makes it "important".

Oh, yes, I like that. `S` is great. How about simply `--session` for the long option?

ahesford commented:
`--session` seems fine to me.

stefonarch commented:
Preliminary test results:
* just one process is fine `labwc -p qterminal` → `exit` → labwc exits.
* the same, opening `lximage-qt` and lowering it, `exit` in qterminal and labwc doesn't exit.
* with `-p lxqt-session` in the start script it doesn't exit labwc when lxqt-session exits
* with a script `start-lxqt` which launches just panel and desktop opening a terminal and killing both labwc doesn't exit

johanmalm commented:
I've changed the command line option to `-S|--session` and added documentation.

Also, description added in `labwc.1.scd`

It works here... if I do the following:
1. `labwc -S qterminal` (running nested)
2. Opening `lximage-qt` in the `qterminal`
3. Minimize `lximage-qt`
4. Exit `qterminal` by pressing close-button in top-right corner
5. Observe that `labwc` terminates

johanmalm commented:
I've just successfully tested the following steps (nested and DRM):

- `labwc -S lxqt-session`
- Open a terminal
- Run `lxqt-leave`
- Click logout
- Observe that compositor terminates.

Consolatis commented on docs/labwc.1.scd:
```diff
@@ -56,6 +56,19 @@ the `--exit` and `--reconfigure` options use.
+*-S, --session* <command>
+	Run command on startup and terminate compositor on exit. This is useful
+	for session management as it allows the session client to terminate
+	labwc be exiting itself. This is a Wayland specific use-case because
```
`by existing`

johanmalm replied:
Thanks. Done.

Consolatis commented on include/common/spawn-primary-client.h:
```diff
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+#ifndef LABWC_SPAWN_PRIMARY_CLIENT_H
+#define LABWC_SPAWN_PRIMARY_CLIENT_H
```
Why the own file and include rather than `common/spawn.{h,c}`? Because of the license?

johanmalm replied:
Because:

1. The license & copyright notice.
2. It's quite a specialist thing with a couple helper functions, so nice to keep them isolated/separated.
3. Makes it clear where those bits have come from for reference for us later (or for future devs) in case the cage approach develops of we want to make changes to it.

stefonarch commented:
I still can't make it work:

* with only `labw -S qterminal` → `lximage-qt & ` → exit terminal labwc doesn't exit
* same with full session 

```
12222:mar 10 08:42:30 archlinux labwc[71328]: lxqt-session: Seding TERM to child  71845
12223:mar 10 08:42:30 archlinux labwc[71328]: lxqt-session: Seding TERM to child  72846
12224-mar 10 08:42:30 archlinux labwc[71328]: lxqt-session: Child process  71406  terminated on signal  15
12225-mar 10 08:42:30 archlinux labwc[71328]: lxqt-session: Child process  71809  terminated on signal  15
12226-mar 10 08:42:30 archlinux labwc[71328]: lxqt-session: Child process  72846  exited with status  0
12227-mar 10 08:42:30 archlinux labwc[71328]: lxqt-session: Child process  71404  exited with status  0
12228-mar 10 08:42:30 archlinux labwc[71262]: 00:00:53.688 [DEBUG] [xwayland/xwm.c:1626] unhandled X11 event: FocusOut (10)
12229-mar 10 08:42:30 archlinux labwc[71262]: 00:00:53.689 [ERROR] [xwayland/xwm.c:1599] xcb error: op ChangeProperty (no minor), code Window (no extension), sequence 323, value 6291459
12230-mar 10 08:42:30 archlinux labwc[71262]: 00:00:53.689 [ERROR] [xwayland/xwm.c:1599] xcb error: op ChangeProperty (no minor), code Window (no extension), sequence 325, value 6291459
12231-mar 10 08:42:30 archlinux labwc[71262]: 00:00:53.689 [ERROR] [xwayland/xwm.c:1599] xcb error: op DeleteProperty (no minor), code Window (no extension), sequence 326, value 6291459
12232-mar 10 08:42:30 archlinux labwc[71941]: Exiting due to channel error.
12233-mar 10 08:42:31 archlinux labwc[71802]: Exiting due to channel error.
12234-mar 10 08:42:31 archlinux labwc[72533]: Exiting due to channel error.
12235-mar 10 08:42:31 archlinux labwc[71961]: Exiting due to channel error.
12236-mar 10 08:42:31 archlinux labwc[72112]: Exiting due to channel error.
12237-mar 10 08:42:31 archlinux labwc[71946]: Exiting due to channel error.
12238-mar 10 08:42:31 archlinux labwc[72473]: Exiting due to channel error.
12239-mar 10 08:42:31 archlinux labwc[71262]: 00:00:53.836 [DEBUG] [types/wlr_compositor.c:692] New wlr_surface 0x578941866640 (res 0x578941898940)
12240-mar 10 08:42:31 archlinux labwc[71262]: 00:00:53.837 [DEBUG] [types/xdg_shell/wlr_xdg_surface.c:388] new xdg_surface 0x57894183ae30 (res 0x578941860210)
12241-mar 10 08:42:31 archlinux labwc[71262]: 00:00:53.838 [DEBUG] [types/wlr_xdg_decoration_v1.c:219] new xdg_toplevel_decoration 0x57894160edc0 (res 0x57894170c060)
12242-mar 10 08:42:31 archlinux labwc[71262]: 00:00:53.873 [DEBUG] [../src/view-impl-common.c:49] [map] identifier=com.github.hluk.copyq, title=CopyQ
12243-mar 10 08:42:31 archlinux labwc[71262]: 00:00:53.956 [DEBUG] [types/wlr_compositor.c:692] New wlr_surface 0x578941865120 (res 0x57894185e750)
12244-mar 10 08:42:41 archlinux labwc[71262]: 00:01:04.023 [INFO] [xwayland/server.c:215] Restarting Xwayland (lazy)
12245-mar 10 08:42:43 archlinux labwc[71262]: 00:01:06.302 [DEBUG] [../src/action.c:679] Handling action 14: ShowMenu
12246-mar 10 08:42:46 archlinux labwc[71262]: 00:01:08.763 [DEBUG] [../src/action.c:679] Handling action 6: Exit
12247-mar 10 08:42:46 archlinux labwc[71262]: 00:01:08.764 [INFO] [../src/config/session.c:131] Updating dbus execution environment
12248-mar 10 08:42:46 archlinux labwc[71815]: wl_display_dispatch: Broken pipe

```

stefonarch commented:
Ok, if I remove/rename `~/.config/autostart`  and my session starts only with swaybg and lxqt-modules it works.

johanmalm commented:
Just demonstrating it on my setup. `lxqt-session` does not issue `labwc -e`, it just closes itself.

[recording.webm](https://github.com/labwc/labwc/assets/6955353/bad72183-cdb6-447a-8d00-c4761e696520)

johanmalm commented:
@Tamaranch  @spl237 Any thoughts from either of you. Would this be a useful feature in conjunction with your respective projects?

spl237 commented:
I don't think it is useful for us, at least not in the way we have implemented wayland sessions at present. We use lightdm to start the wayland sessions directly; we use lxsession for X sessions, but we don't use a session manager for wayland, as lightdm does everything we need.

That's not to say it might not be useful in the future!

stefonarch commented:
@johanmalm It works here 
* with `lxqt-session` as long there are no applications configured  for autostart (in 
`lxqt-config-session`  → autostart) = `~/.config/autostart/` empty
* as long no command with & is issued in the terminal using   `labwc -S qterminal`

Tamaranch commented:
Eh, why not. As it happens, I've just merged a [small port to Wayland](https://gitlab.xfce.org/xfce/xfce4-session/-/merge_requests/49) of xfce4-session, where I use labwc as the default compositor, and more specifically the command `labwc --startup xfce4-session`. So this option could be used instead.

That said, until we develop our own compositor (which probably won't be until Xfce 4.22, i.e. not until next year), and even when we do I think, we don't want to impose any particular compositor. So we need something a little generic to leave the session.

For the moment [I've put this in](https://gitlab.xfce.org/xfce/xfce4-session/-/commit/8e623bc442d44f573c7cb4ea69304dfb6f59b5d7), but it can be considered temporary. As discussed in https://github.com/labwc/labwc/issues/1595, the notion of session on Wayland is a bit nebulous at the moment. Perhaps merging the [session-management protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/18) would help clarify things.

stefonarch commented:
Playing whith my test user I see that it depends on _what_ is set in autostart: I added qterminal, firefox and feathernotes and it exits still fine, so there is some blocking app under my autostart. Nice!

Consolatis commented:
> Hidden eggs?
> 
> ```
> $ labwc -p qterminal
> 00:00:00.001 [../src/action.c:430] Invalid action: showdesktop       
> 00:00:00.002 [../src/action.c:430] Invalid action: ToggleAlwaysTop
> ```

No, those must be defined in your `rc.xml`.

johanmalm commented:
> > `00:00:00.001 [../src/action.c:430] Invalid action: showdesktop`

There is an old PR for this, but not implemented yet.

> > `00:00:00.002 [../src/action.c:430] Invalid action: ToggleAlwaysTop`

Missing `On`?

johanmalm commented:
> For the moment [I've put this in](https://gitlab.xfce.org/xfce/xfce4-session/-/commit/8e623bc442d44f573c7cb4ea69304dfb6f59b5d7), but it can be considered temporary. As discussed in #1595, the notion of session on Wayland is a bit nebulous at the moment. Perhaps merging the [session-management protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/18) would help clarify things.

Thanks for your thoughts and for the the link to that protocol. I wasn't aware of that one. Interesting indeed.
The session living through compositor re-starts would be interesting :smile:
Maybe that points to a two-part session manager with some IPC.
I've not really thought about Compositor re-starts - but it feels pretty deep. Had some discussions on IRC the other day, but I couldn't quite tune in to it.

stefonarch commented:
>No, those must be defined in your rc.xml.

In fact, I use a different location usually, was on old file, `showdesktop` was a script which I removed.

> The session living through compositor re-starts would be interesting 😄

kwin_wayland with Qt  have this feature, but it works only in plasma, have tested it killing kwin, it restores  quite everything.

stefonarch commented:
I cleaned up my autostart folder and now it works like a charm.

Consolatis commented on src/common/spawn-primary-client.c:
```diff
@@ -0,0 +1,119 @@
+	pid_t pid = fork();
+	if (!pid) {
+		sigset_t set;
+		sigemptyset(&set);
+		sigprocmask(SIG_SETMASK, &set, NULL);
+		close(fd[0]);
+		execvp(argv[0], argv);
```
I think this misses some other stuff, namely:
- `restore_nofile_limit()`
- `signal(SIGPIPE, SIG_DFL);`

See `src/common/spawn.c`. IMHO we should try to make this a reusable function so we don't accidentally end up adding something that is only reset at one place but not the other. Also: do we need / want `setsid()` ?

Consolatis commented:
> I cleaned up my autostart folder and now it works like a charm.

Should we try to reproduce the issue and figure out why it wasn't working before?
E.g. do we actually trigger `wl_display_terminate()` or not in this case? It seems we do not which is strange.

stefonarch commented:
> > I cleaned up my autostart folder and now it works like a charm.
> 
> Should we try to reproduce the issue and figure out why it wasn't working before? E.g. do we actually trigger `wl_display_terminate()` or not in this case? It seems we do not which is strange.

I _guess_ it was because I had modified lxqt-modules (needed time ago for hyprland) in autostart. I put everything back except them as they are not needed anymore. I could load the backup  and retest, removing one by one.

stefonarch commented:
> The session living through compositor re-starts would be interesting 😄

Found it again, nested session:

`labwc -C /tmp -s lxqt-panel`
closing the window:
`qt.qpa.wayland: Creating a fake screen in order for Qt not to crash`

stefonarch commented:
@johanmalm I've added in the coming version of `lxqt-labwc-session` a version check:
https://github.com/stefonarch/lxqt-labwc-session/blob/c4b1d71c83eba459462f0c1d0f824c5f71968e25/startlxqtlabwc.in#L99
 Any hope to have this feature in v7.2 ?

johanmalm commented:
> @johanmalm I've added in the coming version of `lxqt-labwc-session` a version check: https://github.com/stefonarch/lxqt-labwc-session/blob/c4b1d71c83eba459462f0c1d0f824c5f71968e25/startlxqtlabwc.in#L99 Any hope to have this feature in v7.2 ?

I'll take a look tomorrow to see if I can respond to outstanding review comments. Should be fine for `0.7.2`

johanmalm commented:
> I think this misses some other stuff, namely:
> 
>     * `restore_nofile_limit()`
>     * `signal(SIGPIPE, SIG_DFL);`
> 
> See `src/common/spawn.c`. IMHO we should try to make this a reusable function so we don't accidentally end up adding something that is only reset at one place but not the other.

I have:
- Rebased against master to enable sharing with the spawn-related code that we added with pipemenus.
- Explored `set_cloexec()` and `reset_signals_and_limits()`
- Used those two in `spawn-primary-clients.c` to deduplicate code and address your (@Consolatis) point above.

> Also: do we need / want `setsid()` ?

Not sure. Might be better to leave `setsid()` out (but I'm guess here) to make it easier to track things. Thoughts from from @stefonarch  welcome on that one.

Consolatis commented on src/common/spawn.c:
```diff
@@ -25,11 +25,20 @@ reset_signals_and_limits(void)
 {
 	int flags = fcntl(fd, F_GETFD);
-	fcntl(fd, F_SETFD, flags | FD_CLOEXEC);
+	if (flags == -1) {
+		wlr_log(WLR_ERROR, "Unable to set the CLOEXEC flag: fnctl failed");
+		return false;
+	}
+	flags = flags | FD_CLOEXEC;
+	if (fcntl(fd, F_SETFD, flags) == -1) {
+		wlr_log(WLR_ERROR, "Unable to set the CLOEXEC flag: fnctl failed");
+		return false;
+	}
```
We could use `wlr_log_errno()` instead, it just applies a `: strerr(errno)` to the end of the log message.

johanmalm replied:
Use of `wlr_log_errno()` implemented.

Consolatis commented on include/common/spawn.h:
```diff
@@ -4,6 +4,8 @@
 
 #include <sys/types.h>
 
+bool set_cloexec(int fd);
```
This should have a `#include <stdbool.h>`

Consolatis commented on src/common/spawn-primary-client.c:
```diff
@@ -0,0 +1,103 @@
+#include "common/spawn-primary-client.h"
+#include "labwc.h"
+
+static int
+sigchld_handler(int fd, uint32_t mask, void *data)
```
Could we call that differntly? This is not a `SIGCHLD` handler but rather a usual fd watcher that watches for `WL_EVENT_HANGUP | WL_EVENT_ERROR`.

So maybe something like `handle_pipe_error()` instead.

I am also not convinced that we should use the exact cage implementation and not just write our own in `common/spawn.c`. IMHO it basically only clutters the src/common/ directory and is almost the same code that we already have in `spawn.c`. Maybe we could simply change the `spawn_async_no_shell()` function to accept an optional `int *err_pipe` argument and then create a small wrapper function that calls `spawn_sync_no_shell()` and sets up the error handler?

Consolatis commented on src/server.c:
```diff
@@ -574,6 +574,11 @@ server_finish(struct server *server)
 
+	int status;
+	if (primary_client_pid) {
+		waitpid(primary_client_pid, &status, 0);
```
This shouldn't be required anymore because we have a global SIGCHLD handler since the pipe menu implementation.

Consolatis commented:
Hm.. played around with this a bit and the cage implementation doesn't really make sense for us: it is based on a pipe which triggers an error once \*all\* of the write ends of the pipe are closed. The issue here is that the session manager may spawn further clients which then also inherit the write end of the pipe. That is likely also the reason for the behavior seen with this PR that sometimes more than the session manager itself needs to close.

To make the cage solution work for us we'd need cooperation of the session manager itself so that it sets cloexec on some fd we supply via environment variable or command line argument and that drastically limits this feature. Or it needs to ensure that \*all\* spawned sub processes that could have inherit the write end of the pipe are terminated before the session manager itself terminates.

I'll experiment a bit with another solution where we only hold the pipe fd open for our direct child which then does a blocking wait for the grandchild and then closes the write end of the pipe. That should then correctly terminate labwc as soon as the session manager itself closes, regardless of how many other childs it spawned. And we'd also not leak the write end of the pipe to the session manager in the first place.

johanmalm commented:
@Consolatis I like that. Makes sense. So, in summary:
- Spawn in a similar way to `spawn_async_no_shell()` but without double fork.
- In sigchld handler, just terminate if pid==primary-client-pid

Consolatis commented:
> So, in summary:
> * Spawn in a similar way to `spawn_async_no_shell()` but without double fork.
> * In sigchld handler, just terminate if pid==primary-client-pid

Exactly. If this is the behavior we want (e.g. terminate labwc as soon as application given by `--session` terminates) I think this is way less convoluted than to deal with a random pipe that gets inherited by every process in the chain.

johanmalm commented:
@stefonarch Fancy taking this for another test-drive?

stefonarch commented:
> @stefonarch Fancy taking this for another test-drive?

Working flawlessly :)

johanmalm commented:
Interactively rebased.

Ready for final review.

Consolatis commented:
Thanks for all the testing :)
